In [2]:
import tensorflow as tf
import matplotlib as mpl

Step 1 - Import the data

Also split the data into training and test sets, also hold out the last 5000 images from training/testing

In [6]:
fashionMnist = tf.keras.datasets.fashion_mnist.load_data()
(X_train_full, y_train_full), (X_test, y_test) = fashionMnist
X_train, y_train = X_train_full[:-5000], y_train_full[:-5000]
X_valid, y_valid = X_train_full[:-5000], y_train_full[:-5000]

Step 2 - Create the neural net structure

get_layer() and get_weights() to see layers and weight

In [7]:
tf.random.set_seed(42)
model = tf.keras.Sequential()
model.add(tf.keras.layers.InputLayer(input_shape=[28, 28]))
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(300, activation="relu"))
model.add(tf.keras.layers.Dense(100, activation="relu"))
model.add(tf.keras.layers.Dense(10, activation="softmax"))

model.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense (Dense)               (None, 300)               235500    
                                                                 
 dense_1 (Dense)             (None, 100)               30100     
                                                                 
 dense_2 (Dense)             (None, 10)                1010      
                                                                 
Total params: 266610 (1.02 MB)
Trainable params: 266610 (1.02 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


Step 3 - Compile the model

In [8]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer = tf.keras.optimizers.SGD(learning_rate=0.01),
              metrics=["accuracy"])

Step 4 - Train and evaluate model

In [9]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_valid, y_valid))

Epoch 1/30


1719/1719 [==============================] - 4s 2ms/step - loss: 2800196666385688952832.0000 - accuracy: 0.0991 - val_loss: 2.3025 - val_accuracy: 0.1000
Epoch 2/30
1719/1719 [==============================] - 3s 2ms/step - loss: 2.3027 - accuracy: 0.0981 - val_loss: 2.3025 - val_accuracy: 0.1001
Epoch 3/30
1719/1719 [==============================] - 3s 2ms/step - loss: 2.3026 - accuracy: 0.0989 - val_loss: 2.3026 - val_accuracy: 0.1006
Epoch 4/30
1719/1719 [==============================] - 3s 2ms/step - loss: 2.3027 - accuracy: 0.0983 - val_loss: 2.3025 - val_accuracy: 0.0999
Epoch 5/30
1719/1719 [==============================] - 3s 2ms/step - loss: 2.3026 - accuracy: 0.0977 - val_loss: 2.3025 - val_accuracy: 0.1006
Epoch 6/30
1719/1719 [==============================] - 3s 2ms/step - loss: 2.3027 - accuracy: 0.0994 - val_loss: 2.3025 - val_accuracy: 0.1009
Epoch 7/30
1719/1719 [==============================] - 3s 2ms/step - loss: 2.3026 - accuracy: 0.1016 - val_loss:

Test one with 30 epochs, learning rate of .01, and 3 layers one with 300 neurons, one with 100 neurons, and one with 10 neurons: loss is 2.3026/2.3027 and an accuracy between .0977 and .1016.

Step 5 - Check for overfitting and retest if needed (STOP HERE BASED ON ASSIGNMENT REQS)

Step 6 - Visualize the data

Step 7 - Use model to make predictions